internal/resolver/delegatingresolver/delegatingresolver.go

```diff
@@ -0,0 +1,241 @@
+/*
+ *
+ * Copyright 2024 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+// Package delegatingresolver implements the default resolver for gRPC.
+// It handles both target URI and proxy address resolution, unless:
+//   - A custom dialer is set using WithContextDialer dialoption.
+//   - Proxy usage is explicitly disabled using WithNoProxy dialoption.
+//   - Client-side resolution is explicitly enforced using WithTargetResolutionEnabled.
+package delegatingresolver
+
+import (
+	"fmt"
+	"net/http"
+	"net/url"
+	"sync"
+
+	"google.golang.org/grpc/grpclog"
+	"google.golang.org/grpc/resolver"
+	"google.golang.org/grpc/serviceconfig"
+)
+
+var (
+	// The following variable will be overwritten in the tests.
+	httpProxyFromEnvironment = http.ProxyFromEnvironment
```

Comment:
httpsProxyFromEnvironment?

Reply:
Changed. Also made the variable exported to be replaced and used in tests.

Reply:
for test you don't have to export because they are part of same package